### PR TITLE
Combined notary witnesses

### PR DESCRIPTION
**Pull request by alexvanin**
Add invoker witness as the third witness. Required for netmap methods checks (https://github.com/nspcc-dev/neofs-contract/issues/154). 

Blocked by neo-go release, and https://github.com/nspcc-dev/neofs-node/pull/971

### CODE REVIEW

File: pkg/morph/client/notary.go
Comment (AnnaShaleva) [inline pkg/morph/client/notary.go:515]: Please, use `c.acc.ScriptHash()` for that.

File: pkg/morph/client/notary.go
Comment (AnnaShaleva) [inline pkg/morph/client/notary.go:601]: Use `c.acc.SignTx(...)` instead of subsequent 10 lines of code. It does exactly the same thing.

File: pkg/morph/event/notary_preparator.go
Comment (AnnaShaleva) [inline pkg/morph/event/notary_preparator.go:342]: s/a/an

**Review by AnnaShaleva — CHANGES_REQUESTED** [review]
I have several questions:
1. The code assumes that third witness (the invoker's one) is *always* present, which means that even when `invokedByAlpha` is true, the invoker signature wil also be attached to the transaction. Question: who is the invoker in case of `invokeByAlpha` is set to true?
2. Initial transaction is signed by the invoker and that's OK for cases when alphabet nodes do not change the transaction hashable part (script, nonce, VUB etc.). But there's the case when alphabet nodes change the initial transaction script, and after that invoker's signature becomes invalid. Is it possible? And if so, then do you handle it somehow?

**Comment by alexvanin** [comment]
> The code assumes that third witness (the invoker's one) is always present, which means that even when invokedByAlpha is true, the invoker signature wil also be attached to the transaction. Question: who is the invoker in case of invokeByAlpha is set to true?

Invoker is going to be an alphabet node, therefore 3rd witness is a regular signature of alphabet node (not multisignature like 2nd witness). I had a choice:
- parse notary requests with variable amount of witnesses (3 for `invokeByAlpha = true` and 4 for `invokeByAlpha = false`),
- add mandatory invoker wintess even if it unnecessary for `invokeByAlpha = true`. 

I thought it will be more consistent to have the one scheme with 4 witnesses for all NeoFS related notary requests. It should not break anything but it is redundant, I agree. Is it bad?

> Initial transaction is signed by the invoker and that's OK for cases when alphabet nodes do not change the transaction hashable part (script, nonce, VUB etc.). But there's the case when alphabet nodes change the initial transaction script, and after that invoker's signature becomes invalid. Is it possible? And if so, then do you handle it somehow?

It is possible. We have one case with `netmap.AddPeer` method (https://github.com/nspcc-dev/neofs-contract/issues/154). To solve it we are going to split this method into two separate methods. _First_ method, invoked by Storage node, only checks witness of invoker. Alphabet node creates new notary request, that invokes _second_ method, which only checks witness of Alphabet.

**Comment by AnnaShaleva** [comment]
> I thought it will be more consistent to have one scheme with 4 witnesses for all NeoFS related notary requests. It should not break anything but it is redundant, I agree. Is it bad?

I also thought about its redundancy, just needed to clarify. I think it's not critical and we may safely use the common scheme with 4 witnesses for all cases.

> It is possible. We have one case with netmap.AddPeer method (nspcc-dev/neofs-contract#154). To solve it we are going to split this method into two separate methods. First method, invoked by Storage node, only checks witness of invoker. Alphabet node creates new notary request, that invokes second method, which only checks witness of Alphabet.

I firstly have a thought that except two different `netmap.AddPeer` methods we also need two different methods for client to construct proper notary signers/witnesses for two different cases. But after the second review, I realised that we don't need to split them on the client side, so everything is OK with the current code.

**Comment by AnnaShaleva** [comment]
>  But after the second review, I realised that we don't need to split them on the client side, so everything is OK with the current code.

Or maybe we still need. Consider the case: storage node constructs notary request with `netmap.AddPeer` call and sends it to the network. Alphabet nodes receive the request and construct their own, it contains 4 signers: proxy + alphabet + invoker + notary. And `invoker` here is an alphabet node (you wrote the same):

> Invoker is going to be an alphabet node, therefore 3rd witness is a regular signature of alphabet node (not multisignature like 2nd witness).

Each alphabet node should construct same main transaction with same set of signers. And here's the problem: alphabet node `A` will put herself as the 3rd signer (because it's a standard signature signer, not multisig); alphabet node `B` will put herself as the 3rd signer and etc. As a result, main transactions generated by alphabet nodes will be different (3rd signer is unique).

So we need to change our signing scheme, otherwise even standard notary requests won't be collected.

**Comment by alexvanin** [comment]
> Each alphabet node should construct same main transaction with same set of signers. And here's the problem: alphabet node A will put herself as the 3rd signer (because it's a standard signature signer, not multisig); alphabet node B will put herself as the 3rd signer and etc. As a result, main transactions generated by alphabet nodes will be different (3rd signer is unique).

Oh, you are right. I guess we have no other choice other than use variable amount of witnesses in notary requests: 4 for `invokeByAlpha = false` and 3 for `invokeByAlpha = true`. But that's okay.

File: pkg/morph/client/notary.go
Comment (alexvanin) [reply pkg/morph/client/notary.go]: Can you give a reference for that in latest neo-go? Couldn't find `ScriptHash()` implementation on account structure.

Comment (AnnaShaleva) [reply pkg/morph/client/notary.go]: Oh, sorry, that's my bad, there's no such API defined on `Account`. There's only `acc.PrivateKey().GetScriptHash()`, but it's better to keep ` hash.Hash160(c.acc.GetVerificationScript())` then.

**Review by cthulhu-rider — APPROVED** [review]

**Comment by alexvanin** [comment]
Merged to unblock subnetwork implementation. Any other witness related changes may be addressed in separate issues and PRs.

**Pull request merged**
